### ISSUE TITLE
Concurrent model weight download

### ISSUE THREAD

**Issue by lmmx** [huggingface/huggingface_hub#1818]
**Is your feature request related to a problem? Please describe.**

I just looked at how HF Hub model weight downloads are implemented and it looks like plain requests :disappointed:

https://github.com/huggingface/huggingface_hub/blob/9b9baa0f41e1c92683c4e54ed9112c3d5ce04ca1/src/huggingface_hub/utils/_http.py#L73

https://github.com/huggingface/huggingface_hub/blob/9b9baa0f41e1c92683c4e54ed9112c3d5ce04ca1/src/huggingface_hub/utils/_http.py#L267

Unless I missed it, there's no concurrency here, despite the common practice of model weights being large and sharded, and therefore a prime target for speedup by concurrent/parallel download.

In Python you can run multiple requests in parallel by looping over the `concurrent.futures`, I implemented it recently in a pretty advanced way and saw 92% faster web requests. I think that in the case of model weights we could expect more on the order of a factor of the number of shards (so for a model in 4 parts, a 4x speedup). Currently it seems to be completely linear.

I expect you could get further speedup (e.g. to speedup a single model weight file, one that's not in parts) by using range requests to split the work between threads. This is more advanced, I wouldn't begin work there.

You use tqdm for progress bars during the download: something I recently discovered was that you can set a parameter [position](https://tqdm.github.io/docs/tqdm/) which will display the progress bar at an offset down the screen, and thus allow you to show progress of multiple files in parallel (if you don't pass this they will 'fight' one another on the screen).

**Describe the solution you'd like**

To get parallel requests I would use a `ThreadPoolExecutor` and loop over the concurrent futures (each future being a "task" that you "submit", e.g. `session.get(model_weight_shard_url)`). I would set a signal handler for SIGINT so that you can cancel the threads (otherwise a thread can end up 'backgrounded' and the download would continue after cancellation, not good in the case of large downloads) and wrap this in a try/except block that either simply raises or outright calls `os._exit` (like `kill` in a shell).

**Describe alternatives you've considered**

The alternative is to use a `ProcessPoolExecutor` but that is not optimal for IO-bound tasks like web requests, you use that for CPU-bound tasks where you're parallelising something that does something computationally strenuous.

**Additional context**

I can provide further code and contribute to a demo/proof of concept if this feature is desirable.

**Comment by Wauplin** [comment]
Hey @lmmx, legitimate concern here! 

Let me share what we are currently doing:
- when downloading a file with `hf_hub_download`, a single `request.get(..., stream=True)` is used => no parallelization by default for single files
- when downloading multiple files with `snapshot_download`, we use `tqdm.contrib.concurrent.thread_map` which is a integration of `ThreadPoolExecutor` in tqdm => meaning parallelization by default for sharded models
- for faster downloads on machines with high bandwidth, we built [hf_transfer](https://github.com/huggingface/hf_transfer) a Rust-based library that download ranges of the file in parallel, for every single file => similar to what you suggested but even faster since it's built in Rust (no GIL) and no need for multiple processes. To enable `hf_transfer` you need to install it (`pip install -U hf_transfer`) and set [`HF_HUB_ENABLE_HF_TRANSFER=1`](https://huggingface.co/docs/huggingface_hub/package_reference/environment_variables#hfhubenablehftransfer) as environment variable. Since it is a library optimized for speed, UX is slightly degraded (no proxies, less error handling, no resumable downloads) so you need to know what you're signing for (that's why it's an opt-in feature). But you can easily expect downloads at 1GB/s if your network and machine allows it!

**Note:** the above is also true for file uploads.

**Comment by Wauplin** [comment]
Doing some cleaning on the repo. I'll close this issue given the response in https://github.com/huggingface/huggingface_hub/issues/1818#issuecomment-1807899136. Please let me know if you have more question and I can reopen.

**Comment by larin92** [comment]
this issue is not completely related to my question, but this is the closest i was able to find.

how about downloading several models from huggingface at once? doing this from several VMs at once does seem to freeze downloads for every VM. each VM tries to download it's own adapter version (hence different models) but all VMs use the same apikey. not sure if providing several/different/unique apikeys would work. is there a way to work around this i'm missing?

**Comment by Wauplin** [comment]
Hi @larin92, to answer your question we don't block concurrent downloads based on user account/HF_TOKEN. So it's ok to use a single one. It would be interesting to know at which step precisely the VMs are freezing (typically doing a ctrl+C and looking at the stacktrace). Also, are the VMs sharing the same cache directory? We have a filelock to prevent downloading twice the same file from the same repo at the same time, meaning that if a VM is downloading a file and another VM tries to download it as well, it will freeze until the first VM has completed. 

(btw if your problem persists, could you open a new issue to discuss this? Will be more convenient to track it)
